# 🤖 Fine-tuning de LLMs com QLoRA Gemma 3 1B

## Bem-vindo/a

Neste notebook você aprenderá a fazer **fine-tuning** de um modelo de linguagem grande (LLM) usando uma técnica chamada **QLoRA**, que permite treinar modelos poderosos mesmo com a GPU gratuita do Google Colab.

Ao final deste laboratório você terá:
1. Entendido a diferença entre modelos **PT** (Pre-Trained) e **IT** (Instruction-Tuned)
2. Treinado seu próprio modelo de geração de código Python
3. Enviado seus pesos treinados para o **HuggingFace**
4. Comparado seu modelo com o original em exemplos reais

---

## 📋 Como usar este notebook

Este notebook está dividido em **3 blocos grandes**. Cada bloco é executado de forma independente para não sobrecarregar a GPU do Colab.

| Bloco | Conteúdo | Ação ao terminar |
|--------|-----------|-----------------|
| **Bloco 1** | PT vs IT: entendendo os modelos | ⚠️ Reiniciar kernel |
| **Bloco 2** | Fine-tuning + Salvar pesos | ⚠️ Reiniciar kernel |
| **Bloco 3** | Comparativo de inferências | ✅ Fim |

> **Tip:** No Google Colab, para reiniciar o kernel vá em `Runtime → Restart session` (ou `Ambiente de execução → Reiniciar sessão`). Isso libera a memória da GPU para o próximo bloco.

---

## ⚙️ Configuração do Google Colab

Antes de começar, certifique-se de que o Colab está usando uma GPU:

1. Vá em **`Runtime → Change runtime type`** (ou `Ambiente de execução → Alterar tipo de ambiente de execução`)
2. Selecione **`T4 GPU`** em Hardware accelerator
3. Clique em **Save**

Você pode verificar que tem GPU executando a célula abaixo:


In [ ]:
# Verificar que temos GPU disponível
import subprocess
result = subprocess.run(['nvidia-smi'], capture_output=True, text=True)
if result.returncode == 0:
    print("GPU detectada:")
    print(result.stdout)
else:
    print("Nenhuma GPU detectada.")

GPU detectada:
Sun May 10 22:19:26 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   38C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+--------------------------------

---

# 🟦 BLOCO 1: PT vs IT - Qual é a diferença?

## Objetivos deste bloco
- Entender o que é um modelo **Pre-Trained (PT)** e como ele gera texto livremente
- Entender o que é um modelo **Instruction-Tuned (IT)** e como ele interpreta instruções
- Conhecer os **tipos de papéis** em uma conversa (system, user, assistant)
- Ver uma **comparação direta** entre PT e IT com o mesmo prompt

---

## 🔑 Passo 1: Criar sua conta no HuggingFace e obter um token

O HuggingFace é a plataforma onde vivem os modelos que vamos usar. Para acessar o **Gemma 3** (o modelo do Google), você precisa criar uma conta e obter um token de acesso.

### Instruções passo a passo:

**1. Criar conta:**
- Vá em [https://huggingface.co/join](https://huggingface.co/join)
- Registre-se com seu e-mail
- Confirme sua conta pelo e-mail de verificação

**2. Obter seu token de acesso (READ):**
- Após fazer login, clique na sua foto de perfil (canto superior direito)
- Vá em **Settings → Access Tokens**
- Clique em **"New token"**
- Dê um nome (ex: `colab-lab`) e selecione o tipo **"Read"**
- Clique em **Generate token**
- **Copie o token** — ele começa com `hf_...`

**3. Aceitar os termos de uso do Gemma 3:**
- Vá em [https://huggingface.co/google/gemma-3-1b-pt](https://huggingface.co/google/gemma-3-1b-pt)
- Aceite os termos de uso do Google clicando no botão que aparece
- Repita em [https://huggingface.co/google/gemma-3-1b-it](https://huggingface.co/google/gemma-3-1b-it)

> **Importante:** Sem aceitar os termos de uso, o modelo será baixado com um erro de acesso negado (401).



---

### 🎯 MILESTONE 1 - Configure seu token

Cole seu token do HuggingFace na célula abaixo (substitua `TU_TOKEN_AQUI`).

> 💡 Em um ambiente de produção, você nunca colocaria o token diretamente no código. Mas para este laboratório está tudo bem, desde que o notebook seja seu.


In [ ]:
import os

# ─────────────────────────────────────────────────────────────────
# 🔑 SUBSTITUA "TU_TOKEN_AQUI" pelo seu token do HuggingFace
#    O token começa com hf_...
#    Exemplo: HF_TOKEN = "hf_abcXYZ123..."
# ─────────────────────────────────────────────────────────────────
# HF_TOKEN = "hf_abcXYZ123"

# Realizando a leitura via COLAB/Secrets
from google.colab import userdata
HF_TOKEN = userdata.get('HF_TOKEN')

# Validação básica do token
if HF_TOKEN == "TU_TOKEN_AQUI" or not HF_TOKEN.startswith("hf_"):
    print("❌ Token inválido. Substitua TU_TOKEN_AQUI pelo seu token do HuggingFace.")
else:
    print("✅ Token configurado corretamente.")

✅ Token configurado corretamente.


---

## 📦 Instalação de bibliotecas

Instalamos as bibliotecas necessárias para este bloco. A flag `-q` suprime a saída verbosa para manter o notebook limpo.


In [ ]:
# Instalamos as bibliotecas principais:
# - bitsandbytes: para quantização de modelos (reduzir memória)
# - peft: Parameter-Efficient Fine-Tuning (LoRA)
# - trl: Transformer Reinforcement Learning (SFTTrainer)
# - accelerate: treinamento distribuído e otimizações
# - datasets: para carregar e processar datasets do HuggingFace
# - transformers: a biblioteca principal do HuggingFace
!pip3 install -q -U bitsandbytes peft trl accelerate datasets transformers

In [ ]:
from transformers import pipeline
import torch

---

## 🧠 O que é um modelo Pre-Trained (PT)?

Um modelo **Pre-Trained (PT)** é um modelo que foi treinado sobre enormes quantidades de texto da internet com um único objetivo: **prever o próximo token** (palavra/símbolo).

```
Entrada:  "O céu é da cor"
Previsão: " azul" (o token mais provável que vem em seguida)
```

Esse processo se chama **pré-treinamento** e dá ao modelo um conhecimento geral da linguagem. Mas o modelo **não sabe seguir instruções**, ele só sabe *continuar texto*.

### Geração livre (free generation)

Quando damos um texto a um modelo PT, ele simplesmente o **completa** como se fosse um documento da internet. O resultado pode ser:
- Continuar a frase de forma natural
- Gerar código se o prompt parecer código
- Gerar texto em qualquer idioma ou formato
- Resultados imprevisíveis e às vezes incoerentes

Vamos ver isso em ação com o **Gemma 3 1B PT**:


In [ ]:
# Carregamos o modelo PT (Pre-Trained) do Gemma 3 1B
# - "text-generation": tarefa de completar/continuar texto
# - device="cuda": usar a GPU
# - dtype=torch.bfloat16: precisão reduzida para economizar memória (bfloat16 é mais estável que float16)

pipe = pipeline("text-generation",
    model="google/gemma-3-1b-pt",   # PT = Pre-Trained (não instruction-tuned)
    device="cuda",
    dtype=torch.bfloat16)

config.json:   0%|          | 0.00/880 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.00G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/215 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/35.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

### Prompt de teste

Observe como o modelo PT responde a "Create a code to sum 3 variables: a, b, c":

> **O que você espera que aconteça?** O modelo NÃO entende que estamos pedindo que ele faça algo. Ele apenas vê texto e o continua.

Execute o código várias vezes.


In [ ]:
# Exemplo de geração livre com o modelo PT
# O modelo simplesmente completa o texto, não "entende" que pedimos código
prompt = "Create a code to sum 3 variables: a, b, c"

output_pt = pipe(prompt, max_new_tokens=50)

print("=" * 60)
print("PROMPT:")
print(prompt)
print()
print("RESPOSTA DO MODELO PT (geração livre):")
print(output_pt[0]["generated_text"])
print("=" * 60)

[transformers] Both `max_new_tokens` (=50) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


PROMPT:
Create a code to sum 3 variables: a, b, c

RESPOSTA DO MODELO PT (geração livre):
Create a code to sum 3 variables: a, b, c

a = 10
b = 20
c = 30

result = a + b + c

print(result)



### 🔍 Observação

Note que o modelo PT pode:
- Continuar o texto de forma literária (escrever mais instruções como se fosse um exercício de programação)
- Gerar código Python diretamente (se durante seu treinamento viu muito código com esse padrão)
- Ou dar uma resposta completamente diferente do que esperávamos

**Isso é geração livre:** o modelo prevê o que é mais provável vir depois daquele texto, sem mais nada.

---

## 💬 O que é um modelo Instruction-Tuned (IT)?

Um modelo **Instruction-Tuned (IT)** é um modelo PT que foi posteriormente treinado com pares de **instrução → resposta**, usando técnicas como **SFT (Supervised Fine-Tuning)** e **RLHF (Reinforcement Learning from Human Feedback)**.

O modelo IT aprendeu a:
1. Reconhecer que o usuário está **pedindo algo**
2. Gerar uma resposta **útil e estruturada**
3. Manter-se dentro de um **papel** (assistente)

### Os 3 papéis de uma conversa

Um modelo IT entende uma conversa estruturada com 3 tipos de mensagens:

| Papel | O que é? | Exemplo |
|-------|----------|---------|
| `system` | Instruções de comportamento para o modelo (quem ele é, o que faz) | "Você é um assistente especialista em Python." |
| `user` | O que o usuário escreve | "Como somo três variáveis?" |
| `assistant` | A resposta do modelo | "Você pode fazer: `result = a + b + c`" |

### O Chat Template

Para que o modelo IT entenda essa estrutura, as mensagens são convertidas em texto com um formato especial chamado **chat template**. Cada modelo tem o seu. Para o Gemma fica assim:

```
<bos><start_of_turn>user
Olá, como vai?<end_of_turn>
<start_of_turn>model
Estou bem, obrigado!<end_of_turn>
```

Esse formato é **crítico**: se você não o usar corretamente ao fazer inferência, o modelo não entenderá bem o contexto.

Vamos ver como funciona com o tokenizer do Gemma IT:


In [ ]:
from transformers import AutoTokenizer

# Carregamos o tokenizer do modelo IT (Instruction-Tuned)
# O tokenizer IT tem o chat_template correto para o Gemma configurado
# Usamos o IT porque ele tem o chat template de instruções configurado
tokenizer = AutoTokenizer.from_pretrained(
    "google/gemma-3-1b-it",   # IT = Instruction-Tuned
    token=HF_TOKEN
)

config.json:   0%|          | 0.00/899 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/35.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

In [ ]:
# Definimos uma mensagem de teste com os 3 papéis
test_messages = [
    # SYSTEM: dizemos ao modelo qual "personalidade" ou "função" ele tem
    {"role": "system", "content": "You are a Python coding assistant. Respond with Python code that solves the task."},

    # USER: a instrução do usuário
    {"role": "user", "content": "Print hello world"},

    # ASSISTANT: a resposta esperada (útil para ver como fica o formato)
    {"role": "assistant", "content": "print('hello world')"},
]

# apply_chat_template converte a lista de mensagens para o formato de texto que o modelo entende
# tokenize=False → nos dá a string de texto (não os IDs de tokens), útil para visualizar
formatted = tokenizer.apply_chat_template(test_messages, tokenize=False)

print("Assim fica a conversa formatada para o modelo:")
print("=" * 60)
print(formatted)
print("=" * 60)
print()
print("Notas:")
print("  <bos>           = Beginning Of Sequence (início do texto)")
print("  <start_of_turn> = Início do turno de um participante")
print("  <end_of_turn>   = Fim do turno")
print("  user / model    = Os papéis (Gemma usa 'model' em vez de 'assistant')")

Assim fica a conversa formatada para o modelo:
<bos><start_of_turn>user
You are a Python coding assistant. Respond with Python code that solves the task.

Print hello world<end_of_turn>
<start_of_turn>model
print('hello world')<end_of_turn>


Notas:
  <bos>           = Beginning Of Sequence (início do texto)
  <start_of_turn> = Início do turno de um participante
  <end_of_turn>   = Fim do turno
  user / model    = Os papéis (Gemma usa 'model' em vez de 'assistant')


### 🔍 O que você acabou de ver

O **chat template** transforma as mensagens JSON em uma string de texto com tokens especiais que o modelo reconhece. É isso que diferencia uma pergunta IT de uma geração PT:

- **PT:** "Create a code to sum..." → o modelo completa como texto
- **IT:** `<start_of_turn>user\nCreate a code...\n<end_of_turn>\n<start_of_turn>model\n` → o modelo sabe que deve *responder* como assistente

Agora carregamos o modelo IT para ver a diferença:


In [ ]:
from transformers import pipeline
import torch

# Carregamos o modelo IT (Instruction-Tuned) do Gemma 3 1B
# Este modelo já foi treinado para seguir instruções
pipe_it = pipeline(
    "text-generation",
    model="google/gemma-3-1b-it",   # IT = Instruction-Tuned
    device="cuda",
    dtype=torch.bfloat16
)

model.safetensors:   0%|          | 0.00/2.00G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/215 [00:00<?, ?B/s]

In [ ]:
# System message: define o papel do assistente
SYSTEM_MESSAGE = (
    "You are a Python coding assistant. "
    "Respond with Python code that solves the task."
)

# Para o modelo IT, passamos as mensagens com a estrutura correta
# O pipeline do HuggingFace aplica o chat_template automaticamente
# quando passamos uma lista de mensagens com este formato
messages = [
    [
        {
            "role": "system",
            "content": [{"type": "text", "text": SYSTEM_MESSAGE},]
        },
        {
            "role": "user",
            "content": [{"type": "text", "text": "Create a code to sum 3 variables: a, b, c "},]
        },
    ],
]

# O modelo IT recebe a instrução e gera uma resposta estruturada
output = pipe_it(messages, max_new_tokens=500)
output

[transformers] Both `max_new_tokens` (=500) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


[[{'generated_text': [{'role': 'system',
     'content': [{'type': 'text',
       'text': 'You are a Python coding assistant. Respond with Python code that solves the task.'}]},
    {'role': 'user',
     'content': [{'type': 'text',
       'text': 'Create a code to sum 3 variables: a, b, c '}]},
    {'role': 'assistant',
     'content': '```python\ndef sum_three_variables(a, b, c):\n  """\n  This function takes three numbers as input and returns their sum.\n\n  Args:\n    a: The first number.\n    b: The second number.\n    c: The third number.\n\n  Returns:\n    The sum of a, b, and c.\n  """\n  return a + b + c\n\n# Example usage:\nnum1 = 10\nnum2 = 20\nnum3 = 30\n\nresult = sum_three_variables(num1, num2, num3)\nprint(f"The sum of {num1}, {num2}, and {num3} is: {result}")\n```\n'}]}]]

In [ ]:
# O pipeline IT retorna uma lista de mensagens, não apenas texto simples
# Extraímos apenas a resposta do assistente para vê-la claramente
generated_text = output[0][0]["generated_text"]  # lista de mensagens
assistant_message = generated_text[-1]            # última mensagem = a do assistant
content = assistant_message["content"]            # a string com o código

print("=" * 60)
print("RESPOSTA DO MODELO IT (instruction-tuned):")
print("=" * 60)
print(content)

RESPOSTA DO MODELO IT (instruction-tuned):
```python
def sum_three_variables(a, b, c):
  """
  This function takes three numbers as input and returns their sum.

  Args:
    a: The first number.
    b: The second number.
    c: The third number.

  Returns:
    The sum of a, b, and c.
  """
  return a + b + c

# Example usage:
num1 = 10
num2 = 20
num3 = 30

result = sum_three_variables(num1, num2, num3)
print(f"The sum of {num1}, {num2}, and {num3} is: {result}")
```



---

## 📊 Comparativo PT vs IT

| | **Modelo PT** | **Modelo IT** |
|---|---|---|
| **Treinamento** | Somente previsão do próximo token | PT + fine-tuning com instruções |
| **Uso** | Geração livre de texto | Seguir instruções, chatbots |
| **Input** | Texto simples | Mensagens estruturadas (system/user/assistant) |
| **Output** | Continuação do texto | Resposta estruturada e útil |
| **Exemplo Gemma** | `google/gemma-3-1b-pt` | `google/gemma-3-1b-it` |

### 💡 Por que isso importa para o fine-tuning?

No **Bloco 2** vamos fazer fine-tuning sobre o modelo **PT** (não o IT). Por quê?

Porque ao fazer fine-tuning a partir do PT:
- **Partimos de um peso geral** para moldar o modelo exatamente para nossa tarefa-alvo
- O modelo PT tem todo o conhecimento base de linguagem e código
- Aplicamos LoRA para adicionar o comportamento de "assistente de Python" sem modificar todos os pesos

Ao final, nosso modelo PT + LoRA deve se comportar de forma similar a um IT, mas especializado em código Python.

---


### 🎯 MILESTONE 2 - Reflexão sobre PT vs IT

Antes de continuar, responda estas perguntas em uma célula de texto:

**Pergunta 1:** Quando você executou o modelo PT com o prompt "Create a code to sum 3 variables: a, b, c", o que ele gerou exatamente? Foi o que você esperava?

**Pergunta 2:** Qual é a principal diferença entre o texto gerado pelo PT e o gerado pelo IT para o mesmo prompt?

**Pergunta 3:** Se o modelo PT apenas prevê o próximo token, como é possível que às vezes ele gere código funcional?

---

## ⚠️ Antes de continuar com o Bloco 2

Reinicie o kernel para liberar a memória da GPU:
1. `Runtime → Restart session` (no Colab)
2. Aguarde o Colab confirmar que o kernel foi reiniciado
3. **Não execute novamente as células deste bloco** — o Bloco 2 tem suas próprias instalações e imports

> A GPU do Colab tem ~15GB de VRAM. Se não reiniciarmos, o modelo do Bloco 1 continuará ocupando memória e o fine-tuning falhará.



**Pergunta 1**: A cada execução, uma resposta diferente foi gerada. Para as 2 primeiras execuções, o modelo apresentou alucinação: "adicionando" semânticas não solicitadas (tais como - randomização, ranges e divisão). Já a 3a execução foi a que mais se aproximou da resposta desejada. Mas, um comportamento esperado, pois o modelo PT não compreende/interpreta o texto do usuário como uma instrução, ele apenas completa o texto com predições de palavras que ele julga serem as mais prováveis.  

1a execução:<br>
The first variable is a random number from 1 to 10. <br> The second variable is a number from 1 to 5000. <br> The third variable is a number from 1 to 100.

2a execução:<br>
Create a code to sum 3 variables: a, b, c, and store the result in a variable: s. <br> Create a code to divide 3 variables: a, b, c, and store the result in a variable: s. <br> Create a code to add 3 variables: a, b

3a execução:<br>
Create a code to sum 3 variables: a, b, c<br>
a = 10<br>
b = 20<br>
c = 30<br>
result = a + b + c<br>
print(result)

**Pergunta 2**: O modelo IT já está preparado (foi especializado) para uma iteração mais específica de Q&A, através de um template de chat (system/user/assistant) por meio do qual ele espera por instruções, para geração de respostas. Uma especialização que proporciona respostas mais assertivas, potencializando, portanto, seu uso!

**Pergunta 3**: Muito provavelmente, em virtude da base de dados utilizada para seu treinamento, onde estruturas de texto semelhantes influenciaram a geração de predições/palavras relacionadas à código.

---

# 🟩 BLOCO 2: Fine-Tuning com QLoRA + Salvar Pesos

## Objetivos deste bloco
- Entender o que é **QLoRA** e por que ele permite fazer fine-tuning com recursos limitados
- Carregar e preparar o dataset **python-codes-25k**
- Treinar o modelo Gemma 3 1B PT em tarefas de código Python
- Salvar os pesos (adapters LoRA) localmente e enviá-los para o HuggingFace

---

## 🧠 O que é QLoRA?

**QLoRA** = **Q**uantization + **Lo**w-**R**ank **A**daptation

É uma combinação de duas técnicas:

### 1. Quantização (Quantization)
Em vez de salvar os pesos do modelo em alta precisão (float32 = 4 bytes por parâmetro), os salvamos em **4 bits** (~0.5 bytes por parâmetro). Isso reduz o uso de memória em ~8x.

```
float32: 1B parâmetros × 4 bytes = 4 GB
int4:    1B parâmetros × 0.5 bytes = 500 MB
```

### 2. LoRA (Low-Rank Adaptation)
Em vez de atualizar todos os pesos do modelo (que seriam centenas de milhões), o LoRA **congela** os pesos originais e adiciona **matrizes pequenas** (adapters) que são treinadas.

```
Peso original (congelado):   W  [4096 × 4096]
Adapters LoRA (treináveis):  A  [4096 × 16]  ×  B  [16 × 4096]
Resultado na inferência: W + A×B
```

O parâmetro `r=16` (rank) controla o tamanho dos adapters. Rank maior = mais capacidade, mas mais memória.

**Vantagem:** Treinamos apenas ~1-5% dos parâmetros totais, com o modelo base quantizado em 4 bits.

---


In [1]:
# Instalamos as bibliotecas (mesmas do Bloco 1)
!pip3 install -q -U bitsandbytes peft trl accelerate datasets transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.7/60.7 MB 15.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 751.0/751.0 kB 24.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 529.0/529.0 kB 26.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.6/10.6 MB 19.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.9/48.9 MB 15.0 MB/s eta 0:00:00


In [2]:
import os
import torch
import transformers
from datasets import load_dataset
from trl import SFTTrainer, SFTConfig
from peft import LoraConfig
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

# ─────────────────────────────────────────────────────────────────
# 🔑 SUBSTITUA pelo seu token do HuggingFace (o mesmo do Bloco 1)
# ─────────────────────────────────────────────────────────────────
# HF_TOKEN = "hf_123"

# Realizando a leitura via COLAB/Secrets
from google.colab import userdata
HF_TOKEN = userdata.get('HF_TOKEN')

# Validação básica
if HF_TOKEN == "TU_TOKEN_AQUI":
    print("❌ Lembre-se de substituir TU_TOKEN_AQUI pelo seu token do HuggingFace")
else:
    print("✅ Token configurado")

✅ Token configurado


---

## 🗄️ Passo 1: Carregar o modelo base com quantização

Carregamos o **Gemma 3 1B PT** (o modelo base sem instruction-tuning) com quantização em 4 bits (QLoRA).

### Configuração BitsAndBytes (quantização)
- `load_in_4bit=True`: carregar os pesos em 4 bits
- `bnb_4bit_quant_type="nf4"`: NormalFloat4, o tipo de quantização mais preciso para pesos de LLMs
- `bnb_4bit_compute_dtype=torch.bfloat16`: embora os pesos estejam em 4 bits, os cálculos são feitos em bfloat16 (mais estável que float16)
- `bnb_4bit_use_double_quant=True`: quantização dupla → comprime ainda mais os parâmetros de quantização


In [3]:
# Identificador do modelo base no HuggingFace Hub
model_id = "google/gemma-3-1b-pt"

# ─────────────────────────────────────────────────────────────────
# Configuração de quantização em 4 bits com BitsAndBytes
# ─────────────────────────────────────────────────────────────────
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,                           # Quantizar para 4 bits para reduzir VRAM
    bnb_4bit_quant_type="nf4",                   # NormalFloat4: melhor tipo de quant. para LLMs
    bnb_4bit_compute_dtype=torch.bfloat16,       # Precisão de cómputo durante forward/backward
    bnb_4bit_use_double_quant=True,              # Quantização dupla: comprime ainda mais
    bnb_4bit_quant_storage=torch.bfloat16,       # Tipo de armazenamento interno dos pesos
)

# Carregar o modelo base (PT) com quantização aplicada
# device_map={"":0} → todo o modelo na GPU 0
model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    dtype=torch.bfloat16,
    device_map={"":0},          # Atribuir todo o modelo à GPU 0
    token=HF_TOKEN
)

# Carregamos o tokenizer do modelo IT para ter o chat_template correto
# O modelo PT não tem chat_template configurado, mas o IT sim
# Durante treinamento e inferência precisamos do mesmo template → usamos o IT
tokenizer = AutoTokenizer.from_pretrained(
    "google/gemma-3-1b-it",
    token=HF_TOKEN
)

print("✅ Modelo e tokenizer carregados corretamente")
print(f"   Modelo: {model_id}")
print(f"   Parâmetros: ~1B")
print(f"   Quantização: 4-bit NF4")

config.json:   0%|          | 0.00/880 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.00G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/215 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/899 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/35.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

✅ Modelo e tokenizer carregados corretamente
   Modelo: google/gemma-3-1b-pt
   Parâmetros: ~1B
   Quantização: 4-bit NF4


---

## 🗂️ Passo 2: Carregar e preparar o dataset

Usamos o dataset **`flytech/python-codes-25k`** do HuggingFace, que contém ~25.000 pares de:
- **instruction**: descrição de uma tarefa de programação em Python
- **output**: código Python que resolve a tarefa

### Divisão do dataset
O ideal para treinar é dividir em 3 partes:
- **Train (80%):** para treinar o modelo
- **Validation (10%):** para monitorar o treinamento
- **Test (10%):** para avaliação final (Bloco 3)

> 💡 Neste caso usaremos apenas o **Train (10%):** pela quantidade de dados e para que o treinamento não demore muito.


In [4]:
from datasets import load_dataset

# Carregar dataset completo do HuggingFace Hub
data = load_dataset("flytech/python-codes-25k")

# ─────────────────────────────────────────────────────────────────
# Divisão do dataset em train / validation / test
# seed=42 garante que a divisão seja sempre a mesma (reprodutibilidade)
# ─────────────────────────────────────────────────────────────────

# 10% para treinamento
split_1 = data["train"].train_test_split(
    test_size=0.90,
    seed=42
)

# Segunda divisão dos 50% restantes: metade validation, metade test (mas não usaremos)
split_2 = split_1["test"].train_test_split(
    test_size=0.50,
    seed=42
)

# Dataset final com as 3 partições (mas não usaremos)
dataset_splits = {
    "train": split_1["train"],       # 10%
    "validation": split_2["train"],  # 40%
    "test": split_2["test"]          # 40%
}

print("✅ Dataset carregado e dividido:")
print(f"   Train:      {len(dataset_splits['train'])} exemplos")
print(f"   Validation: {len(dataset_splits['validation'])} exemplos")
print(f"   Test:       {len(dataset_splits['test'])} exemplos")

README.md: 0.00B [00:00, ?B/s]

python-codes-25k.json:   0%|          | 0.00/26.4M [00:00<?, ?B/s]

python-codes-25k.jsonl:   0%|          | 0.00/25.4M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/49626 [00:00<?, ? examples/s]

✅ Dataset carregado e dividido:
   Train:      4962 exemplos
   Validation: 22332 exemplos
   Test:       22332 exemplos


In [5]:
# Vamos ver como um exemplo do dataset parece antes de processá-lo
# Ele tem dois campos: "instruction" (a tarefa) e "output" (o código solução)
print("Exemplo bruto do dataset (antes de formatar):")
print("=" * 60)
print(dataset_splits["train"][0])

Exemplo bruto do dataset (antes de formatar):
{'output': "```python\nexpression = '2 * (3 + 5]'\nstack = []\nfor char in expression:\n    if char in '({[':\n        stack.append(char)\n    elif char in ')}]':\n        if stack:\n            open_bracket = stack.pop()\n            if not (open_bracket + char in ['()', '{}', '[]']):\n                print(f'Mismatched brackets: {open_bracket} and {char}')\n        else:\n            print(f'Unmatched bracket: {char}')\nif stack:\n    print(f'Unmatched open brackets: {stack}')\n```", 'instruction': "Correct mismatched brackets in '2 * (3 + 5]'", 'input': 'Correcting mismatched brackets in a specific expression...', 'text': "Correct mismatched brackets in '2 * (3 + 5]' Correcting mismatched brackets in a specific expression... ```python\nexpression = '2 * (3 + 5]'\nstack = []\nfor char in expression:\n    if char in '({[':\n        stack.append(char)\n    elif char in ')}]':\n        if stack:\n            open_bracket = stack.pop()\n     

---

## 🔄 Passo 3: Formatar o dataset com a estrutura `messages[]`

O `SFTTrainer` do TRL espera que os dados estejam em formato de **conversa estruturada** com papéis, igual ao que vimos no Bloco 1.

### Por que este formato e não uma string simples?

Antes, o dataset era formatado como uma string bruta (ex: `"Query: ...\nAnswer: ..."`), mas isso tem um problema: o formato de treinamento não coincidia com o de inferência.

Com `messages[]` + `apply_chat_template`:
- ✅ O formato de treinamento é **idêntico** ao de inferência
- ✅ O `SFTTrainer` detecta automaticamente a coluna `messages` e aplica o template
- ✅ Não precisamos de uma `formatting_func` manual


In [6]:
# System message: define o papel do assistente para TODAS as conversações
# Esta mensagem é adicionada no início de cada exemplo do dataset
SYSTEM_MESSAGE = (
    "You are a Python coding assistant. "
    "Respond with Python code that solves the task."
)

def create_conversation(sample):
    """
    Converte um exemplo bruto do dataset {instruction, output}
    para o formato de conversa estruturada {messages: [{role, content}]}
    que o SFTTrainer entende nativamente.
    """
    return {
        "messages": [
            # Turno do sistema: define o comportamento do assistente
            {"role": "system", "content": SYSTEM_MESSAGE},
            # Turno do usuário: a instrução/tarefa
            {"role": "user", "content": sample["instruction"]},
            # Turno do assistente: o código que o modelo deve gerar
            {"role": "assistant", "content": sample["output"]},
        ]
    }

# Aplicar a transformação a cada partição do dataset
# remove_columns elimina as colunas originais (instruction, output)
# e as substitui pela coluna "messages"
for key in dataset_splits:
    dataset_splits[key] = dataset_splits[key].map(
        create_conversation,
        remove_columns=dataset_splits[key].column_names,
    )

print("✅ Dataset formatado com a estrutura messages[]")
print()
print("Exemplo formatado:")
print(dataset_splits["train"][0])

Map:   0%|          | 0/4962 [00:00<?, ? examples/s]

Map:   0%|          | 0/22332 [00:00<?, ? examples/s]

Map:   0%|          | 0/22332 [00:00<?, ? examples/s]

✅ Dataset formatado com a estrutura messages[]

Exemplo formatado:
{'messages': [{'role': 'system', 'content': 'You are a Python coding assistant. Respond with Python code that solves the task.'}, {'role': 'user', 'content': "Correct mismatched brackets in '2 * (3 + 5]'"}, {'role': 'assistant', 'content': "```python\nexpression = '2 * (3 + 5]'\nstack = []\nfor char in expression:\n    if char in '({[':\n        stack.append(char)\n    elif char in ')}]':\n        if stack:\n            open_bracket = stack.pop()\n            if not (open_bracket + char in ['()', '{}', '[]']):\n                print(f'Mismatched brackets: {open_bracket} and {char}')\n        else:\n            print(f'Unmatched bracket: {char}')\nif stack:\n    print(f'Unmatched open brackets: {stack}')\n```"}]}


In [7]:
# Visualizamos como fica o chat template aplicado a um exemplo de teste
# Este é exatamente o texto que o modelo verá durante o treinamento
test_messages = [
    {"role": "system", "content": SYSTEM_MESSAGE},
    {"role": "user", "content": "Print hello world"},
    {"role": "assistant", "content": "print('hello world')"},
]

# tokenize=False → nos dá a string (não os IDs), para podermos vê-la
formatted = tokenizer.apply_chat_template(test_messages, tokenize=False)

print("Chat template aplicado a um exemplo de teste:")
print("=" * 60)
print(formatted)
print("=" * 60)

Chat template aplicado a um exemplo de teste:
<bos><start_of_turn>user
You are a Python coding assistant. Respond with Python code that solves the task.

Print hello world<end_of_turn>
<start_of_turn>model
print('hello world')<end_of_turn>



---

## 📏 Passo 4: Contar tokens por exemplo

Antes de treinar precisamos saber quantos tokens têm os exemplos do dataset. Isso nos permite:

1. **Escolher `max_length`** no trainer: se colocarmos um valor muito pequeno, truncamos exemplos; se colocarmos um muito grande, consumimos mais memória
2. **Entender o dataset**: saber se há exemplos extremamente longos que possam causar problemas

> 💡 Os LLMs trabalham com **tokens**, não com palavras. Uma palavra pode ser 1-3 tokens. O tokenizer do Gemma tem ~256k tokens em seu vocabulário.


In [8]:
import numpy as np
from tqdm import tqdm

def count_tokens_in_split(dataset_split, tokenizer, sample_size=None):
    """
    Mede a distribuição de comprimento (em tokens) dos exemplos do dataset
    após aplicar o chat template.

    Args:
        dataset_split: partição do dataset a medir
        tokenizer: tokenizer do modelo
        sample_size: se especificado, mede apenas uma amostra aleatória

    Returns:
        np.array com a quantidade de tokens de cada exemplo
    """
    if sample_size and sample_size < len(dataset_split):
        # Se for pedida uma amostra, embaralhamos aleatoriamente antes de selecionar
        dataset_split = dataset_split.shuffle(seed=42).select(range(sample_size))

    lengths = []
    for example in tqdm(dataset_split, desc="Contando tokens"):
        # Aplicar o mesmo template que o SFTTrainer usará internamente
        # add_generation_prompt=False porque a resposta do assistant JÁ está incluída
        text = tokenizer.apply_chat_template(
            example["messages"],
            tokenize=False,
            add_generation_prompt=False,
        )
        # Tokenizar o texto para contar seus tokens
        # add_special_tokens=False: o template já inclui os tokens especiais
        tokens = tokenizer(text, add_special_tokens=False)["input_ids"]
        lengths.append(len(tokens))

    return np.array(lengths)

# Medir a distribuição de tokens em todo o split de train
lengths = count_tokens_in_split(
    dataset_splits["train"],
    tokenizer,
    # Descomente a linha abaixo para medir apenas 2000 exemplos (mais rápido):
    sample_size=2000
)

# Exibir estatísticas da distribuição
print(f"\n📊 Distribuição de comprimentos (em tokens):")
print(f"   Exemplos medidos: {len(lengths)}")
print(f"   Mínimo:           {lengths.min()} tokens")
print(f"   Mediana:          {int(np.median(lengths))} tokens")
print(f"   Média:            {int(lengths.mean())} tokens")
print(f"   Máximo:           {lengths.max()} tokens")
print(f"\n→ Usaremos max_length=512 no trainer, que cobre a maioria dos exemplos.")

Contando tokens: 100%|██████████| 2000/2000 [00:04<00:00, 448.49it/s]


📊 Distribuição de comprimentos (em tokens):
   Exemplos medidos: 2000
   Mínimo:           43 tokens
   Mediana:          132 tokens
   Média:            158 tokens
   Máximo:           502 tokens

→ Usaremos max_length=512 no trainer, que cobre a maioria dos exemplos.


---

## ⚙️ Passo 5: Configuração do LoRA

O LoRA adiciona **matrizes de adaptação de baixo rank** às camadas do modelo. Os parâmetros-chave são:

| Parâmetro | Valor | Significado |
|-----------|-------|-------------|
| `r` | 16 | Rank dos adapters. Mais alto = mais capacidade, mas mais memória |
| `lora_alpha` | 16 | Escala os adapters. Quando `alpha == r`, não há escalonamento extra |
| `lora_dropout` | 0.05 | Dropout nos adapters para regularização |
| `target_modules` | "all-linear" | Aplica LoRA a todas as camadas lineares do transformer |
| `modules_to_save` | ["lm_head", "embed_tokens"] | Camadas que são treinadas completamente (não com LoRA) |

> 💡 `modules_to_save` inclui a cabeça do modelo e os embeddings porque são essenciais para a geração de texto e precisam de mais flexibilidade do que o LoRA sozinho oferece.


In [9]:
from peft import LoraConfig

# Configuração dos adapters LoRA
lora_config = LoraConfig(
    r=16,                          # Rank: tamanho das matrizes de adaptação
    lora_alpha=16,                 # Fator de escala (alpha == r → sem escala adicional)
    lora_dropout=0.05,             # Dropout para regularização
    bias="none",                   # Não treinar vieses (bias)
    target_modules="all-linear",   # Aplicar LoRA a TODAS as camadas lineares
    task_type="CAUSAL_LM",         # Tipo de tarefa: Causal Language Modeling (previsão do próximo token)
    modules_to_save=["lm_head", "embed_tokens"],  # Treinar estas camadas completamente
)

print("✅ Configuração LoRA pronta:")
print(f"   Rank (r):        {lora_config.r}")
print(f"   Alpha:           {lora_config.lora_alpha}")
print(f"   Dropout:         {lora_config.lora_dropout}")
print(f"   Target modules:  {lora_config.target_modules}")

✅ Configuração LoRA pronta:
   Rank (r):        16
   Alpha:           16
   Dropout:         0.05
   Target modules:  all-linear


---

## 🏋️ Passo 6: Treinamento

Configuramos e iniciamos o `SFTTrainer` (Supervised Fine-Tuning Trainer) da biblioteca TRL.

### Parâmetros importantes do treinamento

| Parâmetro | Valor | Por quê |
|-----------|-------|---------|
| `max_length` | 512 | Comprimento máximo de sequência (ajustado à mediana do dataset) |
| `packing` | True | Empacota múltiplos exemplos curtos em uma sequência → mais eficiente |
| `num_train_epochs` | 1 | 1 época completa sobre o dataset (suficiente para ver resultados) |
| `per_device_train_batch_size` | 2 | Exemplos por step de GPU |
| `gradient_accumulation_steps` | 4 | Acumula gradientes de 4 steps antes de atualizar → batch efetivo = 2×4 = 8 |
| `gradient_checkpointing` | True | Economiza memória recomputando ativações no backward |
| `learning_rate` | 2e-4 | Taxa de aprendizado para os adapters LoRA |
| `bf16` | True | Treinamento em bfloat16 (mais estável que float16) |
| `save_steps` | 200 | Salva um checkpoint a cada 200 steps |
| `save_total_limit` | 2 | Mantém apenas os 2 checkpoints mais recentes (economiza espaço em disco) |

> ⏱️ **Tempo estimado:** O treinamento leva ~50 minutos em uma GPU T4 do Colab. Não feche a janela do navegador.


In [14]:
dataset_splits

{'train': Dataset({
     features: ['messages'],
     num_rows: 4962
 }),
 'validation': Dataset({
     features: ['messages'],
     num_rows: 22332
 }),
 'test': Dataset({
     features: ['messages'],
     num_rows: 22332
 })}

In [15]:
# Desativar Weights & Biases (logging externo)
os.environ["WANDB_DISABLED"] = "true"

# ─────────────────────────────────────────────────────────────────
# Inicializar o SFTTrainer
# O SFTTrainer detecta automaticamente a coluna "messages" no dataset
# e aplica apply_chat_template internamente — não precisamos de formatting_func
# ─────────────────────────────────────────────────────────────────
trainer = SFTTrainer(
    model=model,                               # Modelo base quantizado
    train_dataset=dataset_splits["train"],     # Dataset de treinamento

    args=SFTConfig(

        output_dir="./outputs-python",         # Pasta onde os checkpoints são salvos
        max_length=512,                        # Comprimento máximo de sequência em tokens
        packing=True,                          # Empacotar exemplos curtos juntos (eficiência)
        num_train_epochs=1,                    # Número de épocas de treinamento

        # Batch e acumulação
        per_device_train_batch_size=2,         # Exemplos processados simultaneamente na GPU
        gradient_accumulation_steps=4,         # Batch efetivo = 2 × 4 = 8 exemplos
        gradient_checkpointing=True,           # Recomputar ativações para economizar VRAM

        # Otimizador
        optim="paged_adamw_8bit",              # AdamW em 8 bits com paginação (menos memória)

        # Learning rate
        warmup_steps=100,                      # Steps de aquecimento antes da LR máxima
        learning_rate=2e-4,                    # Taxa de aprendizado principal

        # Precisão
        fp16=False,                            # Não usar float16 (menos estável)
        bf16=True,                             # Usar bfloat16 (mais estável em A100/T4)

        # Regularização
        max_grad_norm=0.3,                     # Gradient clipping: evita gradientes explosivos

        # Logging
        logging_steps=50,                     # Reportar métricas a cada 50 steps

        # Checkpoints
        save_strategy="steps",                 # Salvar por número de steps
        save_steps=50,                         # Salvar a cada 50 steps
        save_total_limit=2,                    # Manter apenas os 2 checkpoints mais recentes
        save_only_model=True,                  # Salvar apenas o modelo (não o estado do optimizer)

        # Scheduler
        lr_scheduler_type="constant",          # LR constante após o aquecimento

        # Tokenização
        dataset_kwargs={
            "add_special_tokens": False,       # O chat_template já inclui os special tokens
            "append_concat_token": True,       # Separador entre exemplos empacotados
        }
    ),
    peft_config=lora_config,                  # Configuração LoRA
    processing_class=tokenizer,               # Tokenizer para processar o dataset
)

# ─────────────────────────────────────────────────────────────────
# 🚀 INICIAR O TREINAMENTO
# Isso levará ~50 minutos em uma T4 do Google Colab
# Você verá logs a cada 50 steps com a perda (loss) do treinamento
# ─────────────────────────────────────────────────────────────────
trainer.train()

/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:1348: UserWarning: Model has `tie_word_embeddings=True` and a tied layer is part of the adapter, but `ensure_weight_tying` is not set to True. This can lead to complications, for example when merging the adapter or converting your model to formats other than safetensors. Check the discussion here: https://github.com/huggingface/peft/issues/2777
  warnings.warn(msg)


Tokenizing train dataset:   0%|          | 0/4962 [00:00<?, ? examples/s]

Packing train dataset:   0%|          | 0/4962 [00:00<?, ? examples/s]

Step,Training Loss
50,1.171773
100,0.842699
150,0.831414


TrainOutput(global_step=194, training_loss=0.9187889295754973, metrics={'train_runtime': 2923.6135, 'train_samples_per_second': 0.53, 'train_steps_per_second': 0.066, 'total_flos': 4766575041656064.0, 'train_loss': 0.9187889295754973})

---

### 🎯 MILESTONE 3 - Analise o treinamento

Enquanto o treinamento roda (ou quando terminar), observe os logs e responda:

**Pergunta 1:** Qual foi o valor inicial do `loss`? E o valor final? Ele diminuiu?<br>
[Werner] Sim, diminuiu — caiu de ~1.17 para ~0.83, uma redução de cerca de 29%. A queda foi forte entre step 50 e 100, e depois praticamente estabilizou (0.84 → 0.83)

**Pergunta 2:** Quantos steps totais teve o treinamento? (Você vê ao final do training)<br>
[Werner] 194 steps (global_step=194 na linha final do TrainOutput). A barra de progresso [194/194 48:29, Epoch 1/1] confirma — 194 steps em ~48min30s, 1 época completa.

**Pergunta 3 (bonus):** O que aconteceria se treinássemos 3 épocas (`num_train_epochs=3`) em vez de 1? Necessariamente seria melhor? Quanto tempo teoricamente demoraria com os mesmos parâmetros?<br>
[Werner] Considerando 1 época = 194 steps ~ 48 min. 3 épocas = 3*194 = 582 steps ~ 3*48 = 144 min = 2,4 horas. Portanto, demoraria mais, mas difícil afirmar se seria melhor. Considerando que de 100 para 150 steps, o ganho no treinamento foi pontual, aumentar épocas e mais steps, talvez não trará grandes benefícios além de um provável overfitting.

---

## 💾 Passo 7: Inferência local a partir do checkpoint

Antes de enviar o modelo para o HuggingFace, vamos verificar que ele funciona carregando o último checkpoint salvo.


In [16]:
import os
import torch
from peft import PeftModel
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

# ─────────────────────────────────────────────────────────────────
# Buscar o último checkpoint salvo automaticamente
# Os checkpoints são salvos em ./outputs-python/checkpoint-{N}
# onde N é o número do step em que foi salvo
# ─────────────────────────────────────────────────────────────────
checkpoint_dir = "./outputs-python"
checkpoints = [d for d in os.listdir(checkpoint_dir) if d.startswith("checkpoint-")]
last_checkpoint = sorted(checkpoints, key=lambda x: int(x.split("-")[1]))[-1]
checkpoint_path = os.path.join(checkpoint_dir, last_checkpoint)
print(f"✅ Usando checkpoint: {checkpoint_path}")

# ID do modelo base (o mesmo que usamos para treinar)
model_id = "google/gemma-3-1b-pt"

# Mesma configuração de quantização do treinamento
# É importante usar EXATAMENTE a mesma config para compatibilidade
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_storage=torch.bfloat16,
)

# 1. Carregar o modelo BASE (pesos originais do Gemma, congelados)
base_model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    dtype=torch.bfloat16,
    device_map={"": 0},
    token=HF_TOKEN
)

# 2. Carregar os ADAPTERS LoRA treinados sobre o modelo base
# PeftModel combina o base + os adapters em um único objeto
model_finetuned = PeftModel.from_pretrained(base_model, checkpoint_path)

# 3. Tokenizer do modelo IT (para ter o chat_template correto)
tokenizer = AutoTokenizer.from_pretrained("google/gemma-3-1b-it")

# 4. Preparar o modelo para inferência
# use_cache=True: habilita o KV cache para geração mais rápida
# gradient_checkpointing_disable(): não é necessário durante inferência
model_finetuned.config.use_cache = True
model_finetuned.gradient_checkpointing_disable()

print("✅ Modelo fine-tuned carregado para inferência")

✅ Usando checkpoint: ./outputs-python/checkpoint-194


Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

✅ Modelo fine-tuned carregado para inferência


In [17]:
from transformers import pipeline

# System message (deve ser o MESMO que usamos durante o treinamento)
SYSTEM_MESSAGE = (
    "You are a Python coding assistant. "
    "Respond with Python code that solves the task."
)

# Criar pipeline de geração com o modelo fine-tuned
pipe = pipeline(
    "text-generation",
    model=model_finetuned,
    tokenizer=tokenizer,
)

def generate_code(instruction: str, max_new_tokens: int = 200) -> str:
    """
    Gera código Python dado uma instrução.

    A chave é usar apply_chat_template com add_generation_prompt=True:
    Isso adiciona <start_of_turn>model\n ao final do prompt,
    que é o gatilho que o modelo viu durante o treinamento para começar a responder.

    Sem este gatilho, o modelo não saberia quando começar a gerar a resposta.
    """
    # Construir as mensagens da conversa
    messages = [
        {"role": "system", "content": SYSTEM_MESSAGE},
        {"role": "user", "content": instruction},
    ]

    # Aplicar o chat template
    # add_generation_prompt=True → adiciona o gatilho do assistant ao final
    prompt = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=True    # ✅ Adiciona <start_of_turn>model ao final
    )

    # Tokens que indicam o fim da geração
    stop_token_ids = [
        tokenizer.eos_token_id,                               # Token de fim de sequência
        tokenizer.convert_tokens_to_ids("<end_of_turn>")      # Token de fim de turno do Gemma
    ]

    # Gerar a resposta
    outputs = pipe(
        prompt,
        return_full_text=False,        # Retornar apenas o texto NOVO (não o prompt)
        max_new_tokens=max_new_tokens, # Máximo de tokens a gerar
        do_sample=True,                # Amostragem probabilística (não greedy)
        temperature=0.1,               # Temperatura baixa → respostas mais determinísticas e precisas
        top_k=50,                      # Considerar apenas os 50 tokens mais prováveis
        top_p=0.1,                     # Nucleus sampling: top 10% de probabilidade acumulada
        eos_token_id=stop_token_ids,   # Parar ao chegar ao fim de turno ou fim de sequência
        disable_compile=True,          # Desativar torch.compile (compatibilidade com PEFT)
    )

    generated = outputs[0]["generated_text"].strip()
    return generated

# ─────────────────────────────────────────────────────────────────
# Teste inicial do modelo fine-tuned
# ─────────────────────────────────────────────────────────────────
result = generate_code("Create a code to sum 3 variables: a, b, c")
print("INSTRUÇÃO: Create a code to sum 3 variables: a, b, c")
print()
print("RESPOSTA DO MODELO FINE-TUNED:")
print(result)

[transformers] Passing `generation_config` together with generation-related arguments=({'eos_token_id', 'max_new_tokens', 'top_p', 'temperature', 'top_k', 'do_sample', 'disable_compile'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
[transformers] Ignoring clean_up_tokenization_spaces=True for BPE tokenizer GemmaTokenizer. The clean_up_tokenization post-processing step is designed for WordPiece tokenizers and is destructive for BPE (it strips spaces before punctuation). Set clean_up_tokenization_spaces=False to suppress this warning, or set clean_up_tokenization_spaces_for_bpe_even_though_it_will_corrupt_output=True 

INSTRUÇÃO: Create a code to sum 3 variables: a, b, c

RESPOSTA DO MODELO FINE-TUNED:
```python
def sum_variables(a, b, c):
    total = a + b + c
    return total
```


---

## ☁️ Passo 8: Enviar os pesos para o HuggingFace

Agora vamos enviar nossos adapters LoRA para o HuggingFace Hub. Isso tem várias vantagens:
- Os pesos **não se perdem** quando o Colab encerra a sessão
- Você pode compartilhar seu modelo com outros
- No Bloco 3, vamos carregá-lo diretamente do HuggingFace

### Criar um token de escrita (WRITE)

Para enviar modelos você precisa de um token com permissões de **escrita**:

1. Vá em [https://huggingface.co/settings/tokens](https://huggingface.co/settings/tokens)
2. Clique em **"New token"**
3. Dê um nome (ex: `colab-write`)
4. Selecione o tipo **"Write"** (não Read)
5. Clique em **Generate token** e copie-o

> ⚠️ O token de escrita tem mais permissões — não o compartilhe publicamente.

### Nome do seu repositório

O repositório é criado automaticamente com o formato `seu-usuario/nome-do-modelo`.
Troque `TU_USUARIO` pelo seu nome de usuário no HuggingFace.


In [23]:
HF_TOKEN_WRITE

'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'

In [25]:
from huggingface_hub import login

# ─────────────────────────────────────────────────────────────────
# 🔑 SUBSTITUA pelo seu token WRITE do HuggingFace
#    (diferente do token READ que usou antes, precisa de permissões de escrita)
# ─────────────────────────────────────────────────────────────────
# HF_TOKEN_WRITE = "TU_TOKEN_WRITE_AQUI"
from google.colab import userdata
HF_TOKEN_WRITE = userdata.get('HF_TOKEN_WRITE')

# Nome do repositório onde os adapters serão enviados
# ─────────────────────────────────────────────────────────────────
MY_REPO = "werner-cjd/gemma-3-1b-python-coder-v1"  # ← Troque TU_USUARIO
# ─────────────────────────────────────────────────────────────────

# 1. Autenticar no HuggingFace Hub
login(token=HF_TOKEN_WRITE)

# 2. Enviar os adapters LoRA para o Hub
# Isso envia os pesos dos adapters (arquivos adapter_config.json, adapter_model.safetensors)
model_finetuned.push_to_hub(
    MY_REPO,
    private=False,    # False = repositório público (visível para todos)
    token=HF_TOKEN_WRITE # [Werner] Foi necessário adicionar para funcionar
)

# 3. Enviar o tokenizer para o mesmo repositório
# É importante enviar o tokenizer junto com o modelo para que outros possam usá-lo
tokenizer.push_to_hub(
    MY_REPO,
    private=False,
    token=HF_TOKEN_WRITE # [Werner] Foi necessário adicionar para funcionar
)

print(f"✅ Modelo e tokenizer enviados para: https://huggingface.co/{MY_REPO}")

Processing Files (0 / 0)      : |          |  0.00B /  0.00B            

New Data Upload               : |          |  0.00B /  0.00B            

  ...adapter_model.safetensors:   0%|          |  988kB / 1.26GB            

README.md: 0.00B [00:00, ?B/s]

Processing Files (0 / 0)      : |          |  0.00B /  0.00B            

New Data Upload               : |          |  0.00B /  0.00B            

  ...mpurqtgd4e/tokenizer.json:  48%|####7     | 16.0MB / 33.4MB            

✅ Modelo e tokenizer enviados para: https://huggingface.co/werner-cjd/gemma-3-1b-python-coder-v1


---

### 🎯 MILESTONE 4 - Verifique seu modelo no HuggingFace

1. Vá em [https://huggingface.co/TU_USUARIO](https://huggingface.co) (substitua TU_USUARIO pelo seu nome de usuário)
2. Você deve ver o repositório `gemma-3-1b-python-coder` recém-criado
3. Clique no repositório e verifique que os arquivos do modelo aparecem:
   - `adapter_config.json` — configuração dos adapters LoRA
   - `adapter_model.safetensors` — os pesos dos adapters (o resultado do seu treinamento)
   - `tokenizer.json`, `tokenizer_config.json` — o tokenizer

**Compartilhe abaixo a URL do seu modelo**.

> 🎉 Parabéns! Você acabou de treinar e publicar seu primeiro LLM com fine-tuning.<br>
> - https://huggingface.co/werner-cjd/gemma-3-1b-python-coder-v1
---

## ⚠️ Antes de continuar com o Bloco 3

1. **Copie a URL do seu repositório** — você precisará dela no Bloco 3
   - Formato: `TU_USUARIO/gemma-3-1b-python-coder`
2. Reinicie o kernel: `Runtime → Restart session`
3. Não execute nenhuma célula deste bloco novamente



---

# 🟨 BLOCO 3: Comparativo de Modelos — O que o seu aprendeu?

## Objetivos deste bloco
- Carregar seu modelo fine-tuned diretamente do HuggingFace
- Comparar 3 modelos com os mesmos prompts:
  1. 🤖 **Seu modelo fine-tuned** (Gemma PT + LoRA Python)
  2. 🏆 **Gemma 3 1B IT original** (o instruction-tuned do Google)
  3. 🌀 **Gemma 3 1B PT** (geração livre, sem instruction-tuning)
- Explorar as diferenças e limitações de cada um

---

## 📦 Setup inicial


In [1]:
# Instalação de bibliotecas (mesmas dos blocos anteriores)
!pip3 install -q -U bitsandbytes peft trl accelerate datasets transformers

In [2]:
import os
import torch
from peft import PeftModel
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, pipeline

# ─────────────────────────────────────────────────────────────────
# 🔑 Token do HuggingFace (READ, o mesmo do Bloco 1)
# ─────────────────────────────────────────────────────────────────
# HF_TOKEN = "TU_TOKEN_AQUI"
from google.colab import userdata
HF_TOKEN = userdata.get('HF_TOKEN')

# System message: DEVE ser idêntico ao usado durante o treinamento
SYSTEM_MESSAGE = (
    "You are a Python coding assistant. "
    "Respond with Python code that solves the task."
)

print("✅ Configuração pronta")

✅ Configuração pronta


---

## 🤖 Modelo 1: Seu Fine-Tuned (do HuggingFace)

Carregamos diretamente do HuggingFace o modelo que você treinou no Bloco 2.

Substitua `MY_ADAPTER_REPO` pela URL do seu próprio repositório.
Se ainda não tiver o seu pronto, você pode usar o do instrutor como referência: `AleNunezArroyo/gemma-3-1b-python-coder`


In [3]:
# ─────────────────────────────────────────────────────────────────
# OPÇÃO A: Use seu próprio modelo (substitua pelo seu repo)
MY_ADAPTER_REPO = "werner-cjd/gemma-3-1b-python-coder-v1"  # ← Troque isso

# OPÇÃO B: Use o modelo do instrutor como referência
# MY_ADAPTER_REPO = "AleNunezArroyo/gemma-3-1b-python-coder"
# ─────────────────────────────────────────────────────────────────

# Modelo base (o PT, sem instruction-tuning)
model_id = "google/gemma-3-1b-pt"

# Mesma configuração de quantização do treinamento
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_storage=torch.bfloat16,
)

# Carregar o modelo BASE do Gemma
base_model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    dtype=torch.bfloat16,
    device_map={"": 0},
    token=HF_TOKEN
)

# Carregar os adapters LoRA do HuggingFace (os que você enviou no Bloco 2)
model_finetuned = PeftModel.from_pretrained(base_model, MY_ADAPTER_REPO, token=HF_TOKEN)

# Tokenizer do repositório do fine-tuned
tokenizer = AutoTokenizer.from_pretrained(MY_ADAPTER_REPO, token=HF_TOKEN)

# Preparar para inferência
model_finetuned.config.use_cache = True
model_finetuned.gradient_checkpointing_disable()

print(f"✅ Modelo fine-tuned carregado de: {MY_ADAPTER_REPO}")

Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

adapter_config.json: 0.00B [00:00, ?B/s]

/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:1348: UserWarning: Model has `tie_word_embeddings=True` and a tied layer is part of the adapter, but `ensure_weight_tying` is not set to True. This can lead to complications, for example when merging the adapter or converting your model to formats other than safetensors. Check the discussion here: https://github.com/huggingface/peft/issues/2777
  warnings.warn(msg)


adapter_model.safetensors:   0%|          | 0.00/1.26G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/745 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

✅ Modelo fine-tuned carregado de: werner-cjd/gemma-3-1b-python-coder-v1


In [4]:
from transformers import pipeline

# Pipeline com o modelo fine-tuned
pipe_finetuned = pipeline(
    "text-generation",
    model=model_finetuned,
    tokenizer=tokenizer,
)

def generate_finetuned(instruction: str, max_new_tokens: int = 200) -> str:
    """Gera código com o modelo fine-tuned usando o chat template correto."""
    messages = [
        {"role": "system", "content": SYSTEM_MESSAGE},
        {"role": "user", "content": instruction},
    ]

    # Aplicar chat template com add_generation_prompt=True
    # Isso adiciona o token que indica ao modelo que deve começar a gerar a resposta
    prompt = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=True
    )

    stop_token_ids = [
        tokenizer.eos_token_id,
        tokenizer.convert_tokens_to_ids("<end_of_turn>")
    ]

    outputs = pipe_finetuned(
        prompt,
        return_full_text=False,
        max_new_tokens=max_new_tokens,
        do_sample=True,
        temperature=0.1,
        top_k=50,
        top_p=0.1,
        eos_token_id=stop_token_ids,
        disable_compile=True,
    )

    return outputs[0]["generated_text"].strip()

# Teste rápido do modelo fine-tuned
test = generate_finetuned("Create a code to sum 3 variables: a, b, c")
print("Modelo fine-tuned - teste rápido:")
print(test)

[transformers] Passing `generation_config` together with generation-related arguments=({'temperature', 'eos_token_id', 'max_new_tokens', 'top_k', 'do_sample', 'disable_compile', 'top_p'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
[transformers] Ignoring clean_up_tokenization_spaces=True for BPE tokenizer GemmaTokenizer. The clean_up_tokenization post-processing step is designed for WordPiece tokenizers and is destructive for BPE (it strips spaces before punctuation). Set clean_up_tokenization_spaces=False to suppress this warning, or set clean_up_tokenization_spaces_for_bpe_even_though_it_will_corrupt_output=True 

Modelo fine-tuned - teste rápido:
```python
def sum_variables(a, b, c):
    total = a + b + c
    return total
```


---

## 🏆 Modelo 2: Gemma 3 1B IT Original

Carregamos o modelo **Instruction-Tuned oficial do Google** (o mesmo que você viu no Bloco 1). Esta é a referência — o modelo que o Google treinou com milhares de exemplos e RLHF.

> 💡 Este modelo pode responder a qualquer tipo de instrução, não apenas código Python.


In [5]:
from transformers import pipeline
import torch

# Carregar o modelo IT oficial do Google em bfloat16
# Não usamos quantização aqui para manter a qualidade máxima
# Se houver problemas de memória, adicione: dtype=torch.bfloat16
pipe_gemma_it = pipeline("text-generation",
    model="google/gemma-3-1b-it",
    device="cuda",
    dtype=torch.bfloat16)

print("✅ Gemma 3 1B IT oficial carregado")

model.safetensors:   0%|          | 0.00/2.00G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/215 [00:00<?, ?B/s]

✅ Gemma 3 1B IT oficial carregado


In [6]:
def generate_gemma_it(instruction: str, max_new_tokens: int = 500) -> str:
    """
    Gera resposta com o modelo IT oficial do Google.
    Nota: o pipeline do IT retorna uma lista de mensagens (não texto simples),
    por isso extraímos a última mensagem (a do assistant).
    """
    # O pipeline IT do Gemma espera as mensagens com listas de content objects
    messages = [
        [
            {
                "role": "system",
                "content": [{"type": "text", "text": SYSTEM_MESSAGE},]
            },
            {
                "role": "user",
                "content": [{"type": "text", "text": instruction},]
            },
        ],
    ]

    output = pipe_gemma_it(messages, max_new_tokens=max_new_tokens)

    # O pipeline IT retorna uma lista de mensagens completa
    # O último elemento é a resposta do assistant
    generated_text = output[0][0]["generated_text"]  # lista de mensagens
    assistant_message = generated_text[-1]            # último = assistant
    content = assistant_message["content"]            # string com o código

    return content

# Teste rápido do IT oficial
test_it = generate_gemma_it("Create a code to sum 3 variables: a, b, c")
print("Gemma IT oficial - teste rápido:")
print(test_it)

[transformers] Passing `generation_config` together with generation-related arguments=({'max_new_tokens'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
[transformers] Both `max_new_tokens` (=500) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Gemma IT oficial - teste rápido:
```python
def sum_three_variables(a, b, c):
  """
  This function takes three numbers as input and returns their sum.

  Args:
    a: The first number.
    b: The second number.
    c: The third number.

  Returns:
    The sum of a, b, and c.
  """
  total = a + b + c
  return total

# Example usage:
num1 = 10
num2 = 20
num3 = 30

result = sum_three_variables(num1, num2, num3)
print(f"The sum of {num1}, {num2}, and {num3} is: {result}")
```



---

## 🌀 Modelo 3: Gemma 3 1B PT (Geração Livre)

Por completude, também incluímos o modelo PT no comparativo. Lembre-se do Bloco 1 que este modelo simplesmente **continua o texto** sem entender que estamos fazendo uma instrução.


In [7]:
# Carregar o modelo PT para geração livre
# Usado apenas como referência — não entende instruções
pipe_pt = pipeline("text-generation",
    model="google/gemma-3-1b-pt",
    device="cuda",
    torch_dtype=torch.bfloat16)

print("✅ Gemma 3 1B PT carregado para geração livre")

[transformers] `torch_dtype` is deprecated! Use `dtype` instead!


Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/35.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

✅ Gemma 3 1B PT carregado para geração livre


---

## 🧪 Comparativo com exemplos do domínio

Usamos instruções similares às do dataset de treinamento (`flytech/python-codes-25k`) para ver se o modelo aprendeu a resolver tarefas de código Python.

### Exemplos de teste

Testamos com 3 instruções de diferentes níveis de dificuldade:
- **Básica:** soma de variáveis (vista durante o treinamento)
- **Intermediária:** manipulação de listas
- **Avançada:** algo mais estruturado

> 📌 Observe: seu modelo fine-tuned gera código Python válido? É comparável ao IT oficial? O PT dá algo útil?


In [8]:
# ─────────────────────────────────────────────────────────────────
# Exemplos de teste — similares ao domínio do dataset de treinamento
# ─────────────────────────────────────────────────────────────────
test_instructions = [
    "Create a code to sum 3 variables: a, b, c",
    "Write a Python function that takes a list of numbers and returns the maximum value",
    "Write a function that checks if a string is a palindrome",
]

# Executar os 3 modelos sobre cada instrução e exibir resultados comparativos
for i, instruction in enumerate(test_instructions):
    print("=" * 70)
    print(f"📝 EXEMPLO {i+1}: {instruction}")
    print("=" * 70)

    print("\n🤖 SEU MODELO FINE-TUNED:")
    print("-" * 40)
    result_ft = generate_finetuned(instruction, max_new_tokens=200)
    print(result_ft)

    print("\n🏆 GEMMA IT OFICIAL (Google):")
    print("-" * 40)
    result_it = generate_gemma_it(instruction, max_new_tokens=200)
    print(result_it)

    print("\n🌀 GEMMA PT (Geração livre — sem instruction tuning):")
    print("-" * 40)
    result_pt = pipe_pt(instruction, max_new_tokens=100)[0]["generated_text"]
    print(result_pt)

    print()

[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


📝 EXEMPLO 1: Create a code to sum 3 variables: a, b, c

🤖 SEU MODELO FINE-TUNED:
----------------------------------------


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
def sum_variables(a, b, c):
    total = a + b + c
    return total
```

🏆 GEMMA IT OFICIAL (Google):
----------------------------------------


[transformers] Both `max_new_tokens` (=100) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
def sum_three_variables(a, b, c):
  """
  This function takes three numbers as input and returns their sum.

  Args:
    a: The first number.
    b: The second number.
    c: The third number.

  Returns:
    The sum of the three numbers.
  """
  return a + b + c

# Example usage:
num1 = 10
num2 = 20
num3 = 30

result = sum_three_variables(num1, num2, num3)
print(f"The sum of {num1}, {num2}, and {num3} is: {result}") 
```


🌀 GEMMA PT (Geração livre — sem instruction tuning):
----------------------------------------


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Create a code to sum 3 variables: a, b, c

Create a code to get the value of a, b, c

Create a code to print a, b, c

Create a code to get the value of a, b, c

Create a code to get the sum of a, b, c

Create a code to print a, b, c

Create a code to get the sum of a, b, c

Create a code to print a, b, c

Create a code to get the sum of a,

📝 EXEMPLO 2: Write a Python function that takes a list of numbers and returns the maximum value

🤖 SEU MODELO FINE-TUNED:
----------------------------------------


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
def max_value(numbers):
    max_value = 0
    for i in numbers:
        if i > max_value:
            max_value = i
    return max_value

numbers = [1, 2, 3, 4, 5]
print(max_value(numbers))
```

🏆 GEMMA IT OFICIAL (Google):
----------------------------------------


[transformers] Both `max_new_tokens` (=100) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
def find_maximum(numbers):
  """
  Finds the maximum value in a list of numbers.

  Args:
    numbers: A list of numbers.

  Returns:
    The maximum value in the list.  Returns None if the list is empty.
  """
  if not numbers:
    return None  # Handle empty list case

  maximum = numbers[0]
  for number in numbers:
    if number > maximum:
      maximum = number
  return maximum

if __name__ == '__main__':
  # Example usage:
  numbers1 = [1, 5, 2, 8, 3]
  max_value1 = find_maximum(numbers1)
  print(f"The maximum value in {numbers1} is: {max_value1}")  # Output: 8

  numbers2 = [-1, -5, -

🌀 GEMMA PT (Geração livre — sem instruction tuning):
----------------------------------------


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Write a Python function that takes a list of numbers and returns the maximum value in the list. The function should return the largest number in the list if the list is empty.

<code>def max_value(numbers):
    # Your code here
    return max(numbers)
</code>

You can use the <code>max</code> function of the <code>math</code> module in Python to return the largest value in a list.

<code>import math

def max_value(numbers):
    max_value = 0
    for number in numbers:


📝 EXEMPLO 3: Write a function that checks if a string is a palindrome

🤖 SEU MODELO FINE-TUNED:
----------------------------------------


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
def is_palindrome(s):
    # Convert the string to lowercase
    s = s.lower()
    # Convert the string to an integer
    s = int(s)
    # Convert the integer to a string
    s = str(s)
    # Check if the string is a palindrome
    if s == s[::-1]:
        return True
    else:
        return False
```

🏆 GEMMA IT OFICIAL (Google):
----------------------------------------


[transformers] Both `max_new_tokens` (=100) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
def is_palindrome(text):
  """
  Checks if a string is a palindrome (reads the same forwards and backward).

  Args:
    text: The string to check.

  Returns:
    True if the string is a palindrome, False otherwise.
  """
  processed_text = ''.join(filter(str.isalnum, text)).lower()  # Remove non-alphanumeric characters and convert to lowercase
  return processed_text == processed_text[::-1]

# Example usage:
string1 = "racecar"
string2 = "A man, a plan, a canal: Panama"
string3 = "hello"

print(f"'{string1}' is a palindrome: {is_palindrome(string1)}")
print(f"'{string2}' is a palindrome: {is_palindrome(string2)}")
print(f"'{string3}' is a palindrome:

🌀 GEMMA PT (Geração livre — sem instruction tuning):
----------------------------------------
Write a function that checks if a string is a palindrome and returns true if it is, false otherwise.

Write a function that checks if a number is odd or even.

Write a function that checks if a number is an even number and return true

---

### 🎯 MILESTONE 5 - Experimente e reflita

Agora é a sua vez de explorar. Modifique a célula abaixo para testar com suas próprias instruções:


In [9]:
# ─────────────────────────────────────────────────────────────────
# 🧪 ZONA DE EXPERIMENTAÇÃO LIVRE
# Troque estas instruções pelas que você quiser testar
# ─────────────────────────────────────────────────────────────────

my_instructions = [
    # Instrução 1: algo relacionado ao dataset (similar ao treinamento)
    "Write a Python function to verify the size distribution of each feature (instruction, output) in the training dataset",

    # Instrução 2: algo fora do domínio do Python (como o modelo responde?)
    "Explain what overfitting is in the context of machine learning (in simple words)",

    # Instrução 3: coloque aqui algo de sua escolha 🎯
    "I am studying Agentic AI. What is its impact on a Computer Science career?"
]

print("Testando com minhas próprias instruções...")
print()

for instruction in my_instructions:
    print("=" * 70)
    print(f"📝 INSTRUÇÃO: {instruction}")
    print()

    print("🤖 MEU MODELO FINE-TUNED:")
    result_ft = generate_finetuned(instruction, max_new_tokens=200)
    print(result_ft)

    print()
    print("🏆 GEMMA IT OFICIAL:")
    result_it = generate_gemma_it(instruction, max_new_tokens=200)
    print(result_it)
    print()

[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Testando com minhas próprias instruções...

📝 INSTRUÇÃO: Write a Python function to verify the size distribution of each feature (instruction, output) in the training dataset

🤖 MEU MODELO FINE-TUNED:


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
def verify_size_distribution(X, y):
    # Get the size of each feature
    features = [len(X[i]) for i in X]
    # Get the size of the training dataset
    training_size = sum(features)
    # Compare the size of the training dataset to the size of the feature
    if training_size != features:
        return False
    # Return True if the size of the training dataset is greater than the size of the feature
    return True
```

🏆 GEMMA IT OFICIAL:


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


```python
import numpy as np
import pandas as pd

def verify_feature_distribution(data):
  """
  Verifies the size distribution of each feature (instruction, output) 
  in the training dataset.

  Args:
    data: A pandas DataFrame containing the training dataset. 
          It should have columns 'instruction' and 'output'.

  Returns:
    A pandas DataFrame with the feature distribution (instruction, output).
    Returns None if the input data is invalid.
  """

  if not isinstance(data, pd.DataFrame):
    print("Error: Input must be a pandas DataFrame.")
    return None

  if 'instruction' not in data.columns or 'output' not in data.columns:
    print("Error: DataFrame must contain 'instruction' and 'output' columns.")
    return None

  try:
    feature_counts = data['instruction'].value_counts()

📝 INSTRUÇÃO: Explain what overfitting is in the context of machine learning (in simple words)

🤖 MEU MODELO FINE-TUNED:


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


In machine learning, overfitting occurs when a model is trained on a large dataset and then used to make predictions on a smaller dataset. This happens because the model has been trained on the training data, which may not accurately represent the distribution of the data in the real world. When the model is used to make predictions on the test data, it may not perform as well as it did on the training data, leading to poor performance on the test data. This is why it is important to use a good model selection method to avoid overfitting.

🏆 GEMMA IT OFICIAL:


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Okay, let's tackle this!

Here's the Python code to solve the task, followed by an explanation of overfitting.

```python
def solve_problem():
    """
    This function takes a list of numbers as input and returns their average.
    """
    numbers = [1, 2, 3, 4, 5]
    average = sum(numbers) / len(numbers)
    return average

# Example usage:
result = solve_problem()
print(f"The average of the numbers is: {result}")
```

**Explanation of Overfitting**

Okay, let's break down what overfitting is in machine learning.  It's a really important concept to understand.

**In simple terms:**

Imagine you're teaching a computer to recognize pictures of cats. You show it *thousands* of pictures of cats – fluffy, tabby, black, orange, all sorts!  The

📝 INSTRUÇÃO: I am studying Agentic AI. What is its impact on a Computer Science career?

🤖 MEU MODELO FINE-TUNED:


[transformers] Both `max_new_tokens` (=200) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Agentic AI is a field of Artificial Intelligence that focuses on the development of autonomous and self-learning machines. It aims to create machines that can think, feel, and learn like humans, and is considered to be one of the most important areas of AI research.

🏆 GEMMA IT OFICIAL:
Okay, let's tackle the impact of Agentic AI on a Computer Science career. It's a *massive* and rapidly evolving field, so understanding its implications is crucial. Here's a breakdown, presented with Python code and explanations, covering the key areas:

```python
def analyze_agentic_ai_impact():
    """
    Analyzes the impact of Agentic AI on a Computer Science career.
    """

    print("Agentic AI is poised to significantly reshape the landscape of Computer Science.")
    print("Here's a look at its key impacts:\n")

    print("1. **Increased Demand for Specialized Skills:**")
    print("   - Traditional CS roles are becoming less focused on 'routine' programming.  ")
    print("   -  We'll need exp

---

## 🏁 Reflexão final

Responda estas perguntas ao terminar o laboratório:

### Sobre os resultados

**Pergunta 1:** Em que tipo de instruções seu modelo fine-tuned superou (ou igualou) o Gemma IT oficial? Em quais foi pior?<br>
[Werner] (1) FT gerou um código simples, sem documentação e com um bug de lógica (comparando training_size (int) != features (list)); enquanto o Gemma já apresentou uma solução mais completa com docstring, validação de tipo e pandas. (2) FT gerou uma resposta textual, porém com alucionações: conceitos que não descrevem o significado de overfitting, tais como: tamanho relativo dos datasets, não representativade dos dados reais; enquanto o Gemma injetou código Python antes de responder, prejudicando a compreensão pelo usuário. (3) Ambos apresentaram uma resposta condizente, apesar do Gemma (assim como fez na instrução 2) ter injetado código como resposta para o usuário.

**Pergunta 2:** O que aconteceu quando você testou uma instrução que **não é de código Python** (ex: "Explain machine learning")? Esse comportamento faz sentido?<br>
[Werner] O modelo FT respondeu texto, sem forçar ou injetar código! Como o modelo base foi o Gemma PT, o FT deve ter preservado sua capacidade de responder de forma textual.

**Pergunta 3:** O modelo PT gerou algo para as mesmas instruções. Foi útil? Como se compara com os modelos IT?<br>
[Werner] Não explorei nesta comparação, mas observando a execução anterior, respostas foram geradas, mas muito superficiais ou inacabadas, exigindo retrabalho e correção.

### Sobre o processo

**Pergunta 4:** Se você quisesse melhorar os resultados do seu modelo fine-tuned sem alterar o dataset, quais hiperparâmetros modificaria primeiro e por quê?<br>
[Werner] Aumentar o número de épocas para experimentação, assim como o Learning Rate e LoRA rank (r) para 16 ou 32, podendo dar uma maior capacidade de precisão, sem impactar muito o consumo de memória. E aumentaria de 200 para 512 o tamanho máximo das respostas.

**Pergunta 5:** Treinamos com 10% do dataset (em vez de 80%) para economizar tempo. Que efeito você acha que teria usar o dataset completo na qualidade das respostas?<br>
[Werner] Os resultados seriam melhores, pois com mais samples, mais padrões de estrutura - docstrings, listas, códigos - estariam disponíveis para aprendizado, possibilitando inclusive a redução de overfitting, pois o memorizar de exemplos seria menos provável de ocorrer.

*Laboratório de fine-tuning concluído — Gemma 3 1B QLoRA*

